naturalspeech2_pytorch/aligner.py

```diff
@@ -0,0 +1,145 @@
+from typing import Tuple
+import torch
+from torch import nn
+import torch.nn.functional as F
+from einops import rearrange
+import numpy as np
+from numba import jit, prange
+
+class AlignerNet(torch.nn.Module):
+    """alignment model https://arxiv.org/pdf/2108.10447.pdf """
+    def __init__(
+        self,
+        dim_in=80,
+        dim_hidden=512,
+        attn_channels=80,
+        temperature=0.0005,
+    ):
+        super().__init__()
+        self.temperature = temperature
+        self.softmax = torch.nn.Softmax(dim=3)
+        self.log_softmax = torch.nn.LogSoftmax(dim=3)
+
+        self.key_layers = nn.ModuleList([
+            nn.Conv1d(
+                dim_hidden,
+                dim_hidden * 2,
+                kernel_size=3,
+                padding=1,
+                bias=True,
+            ),
+            nn.ReLU(inplace=True),
+            nn.Conv1d(dim_hidden * 2, attn_channels, kernel_size=1, padding=0, bias=True)
+        ])
+
+        self.query_layers = nn.ModuleList([
+            nn.Conv1d(
+                dim_in,
+                dim_in * 2,
+                kernel_size=3,
+                padding=1,
+                bias=True,
+            ),
+            nn.ReLU(inplace=True),
+            nn.Conv1d(dim_in * 2, dim_in, kernel_size=1, padding=0, bias=True),
+            nn.ReLU(inplace=True),
+            nn.Conv1d(dim_in, attn_channels, kernel_size=1, padding=0, bias=True)
+        ])
+
+    def forward(self, queries: torch.Tensor, keys: torch.Tensor, mask: torch.Tensor = None):
```

Comment:
just my thinking out loud, but this is basically attention with the square of the euclidean distance

Reply:
ahh i see, basically creating an alignment matrix for ctc loss got it